docs/doc/60-contributing/03-rfcs/20221105-infer-schema.md

```diff
@@ -0,0 +1,160 @@
+
+For example:
+
+```sql
+CREATE TABLE test BY (
+    INFER 's3://mybucket/data.csv' FILE_FORMAT = ( TYPE = CSV )
+);
+```
+
+Users can also make some changes to the output of `INFER`:
+
+```sql
+CREATE TABLE test BY (
+    SELECT UPPER(COLUMN_NAME), TYPE, TRUE from (
+        INFER 's3://mybucket/data.csv' FILE_FORMAT = ( TYPE = CSV )
+    )
+);
+```
+
+## Reference-level explanation
+
+We will adopt the `infer_schema` feature from `arrow2` to make this possible. Take `csv` as an example:
+
+```rust
+pub fn infer_schema<R: Read + Seek, F: Fn(&[u8]) -> DataType>(
+    reader: &mut Reader<R>,
+    max_rows: Option<usize>,
+    has_header: bool,
+    infer: &F
+) -> Result<(Vec<Field>, usize)>
+```
+
+We will then convert the `Field` into data types that the databend supports.
+
+## Drawbacks
+
+None.
+
+## Rationale and alternatives
+
+None.
+
+## Prior art
+
+### Snowflake
+
+Snowflake has `infer_schema` functions:
+
+```sql
+select *
+  from table(
+    infer_schema(
+      location=>'@mystage'
+      , file_format=>'my_parquet_format'
+      )
+    );
+
++-------------+---------+----------+---------------------+--------------------------+----------+
+| COLUMN_NAME | TYPE    | NULLABLE | EXPRESSION          | FILENAMES                | ORDER_ID |
+|-------------+---------+----------+---------------------+--------------------------|----------+
+| continent   | TEXT    | True     | $1:continent::TEXT  | geography/cities.parquet | 0        |
+| country     | VARIANT | True     | $1:country::VARIANT | geography/cities.parquet | 1        |
+| COUNTRY     | VARIANT | True     | $1:COUNTRY::VARIANT | geography/cities.parquet | 2        |
++-------------+---------+----------+---------------------+--------------------------+----------+
+```
+
+Snowflake's `CREATE TABLE` support `USING TEMPLATE` to accept the output from `infer_schema`:
+
+```sql
+create table mytable
+  using template (
+    select array_agg(object_construct(*))
+    within group (order by order_id)
+      from table(
+        infer_schema(
+          location=>'@mystage',
+          file_format=>'my_parquet_format'
+        )
+      ));
+```
+
+## Unresolved questions
+
+None.
+
+## Future possibilities
+
+### Generate Create Table Statement from INFER
+
+It's possible to generate a `CREATE TABLE` statement from `INFER` in the future, so users can modify the statement in manully instead of by SQL.
+
+For example:
+
+```sql
+SOME MAGIC SQL
+
++--------------------------------------------------+
+| STATEMENT                                        |
+|--------------------------------------------------+
+| CREATE TABLE table_name (name VARCHAR, age INT); |
++--------------------------------------------------+
+```
+
+### SELECT FROM external location
+
+With `INFER` support, we can build a temperary table schema to read data from external location.
+
+```sql
+SELECT * from @my_stage;
+
+SELECT * from 's3://my_bucket/data.csv';
```

Comment:
Cool, under this case, need a new catalog: StageCatalog, and `StageCatalog.schema()` will return the INFER schema?

Reply:
The details still need discussion, but the core idea is right: we can reuse `INFER` internally.